### PR TITLE
dev/core#2243 Add created_date column to the civicrm_note table

### PR DESCRIPTION
**Pull request by jitendrapurohit**
Overview
----------------------------------------
Add created_date column to the civicrm_note table

https://lab.civicrm.org/dev/core/-/issues/2243

Before
----------------------------------------
No date field to know when the note was added. Also, there's no UI field to control/update the date manually.

After
----------------------------------------
Date field added.

image

Is available on view -

image

Comments
----------------------------------------
API test added.

### CODE REVIEW

**Comment by seamuslee001** [comment]
@jitendrapurohit can you fix the style issues https://test.civicrm.org/job/CiviCRM-Core-PR/39602/checkstyleResult/new/

**Comment by sunilpawar** [comment]
@jitendrapurohit Tested the created date functionality, its working fine.

**Comment by mattwire** [comment]
@jitendrapurohit Looks like test fail related? `CRM_Dedupe_BAO_RuleGroupTest::testSupportedFields()`

**Comment by jitendrapurohit** [comment]
Yep. Test build is passing now 👍

**Comment by eileenmcnaughton** [comment]
@jitendrapurohit I agree with  adding this field - but we have a pretty strong convention that the field created_date is a DB populated field normally and it it would normally hold the timestamp - I just added a note and having chosen not to enter the field it winds up empty - which is really against our convention.

I guess it's OK to expose if for override but I don't think it should be OK for it to be blank - @mattwire - @seamuslee001 what do you think here?

**Comment by jitendrapurohit** [comment]
ok, agree with the naming of created_date. Probably, we should -

- Rename `created_date` to `note_date`. So it is just the date attached to the note, not necessarily the date when it was created.
- On create, if it is left empty, it defaults to today's date. No extra handling on edit form.
- Write an upgrade sql after ALTER stmt -`UPDATE civicrm_note SET note_date = modified_date` so that no NULL entries are existing.

Does the above make sense? @eileenmcnaughton @sunilpawar @mattwire

**Comment by eileenmcnaughton** [comment]
@sunilpawar so we chatted on chat & I think the agreement (with @seamuslee001 & @kcristiano ) was

1) yes- change the name to note_date if you are going to expose it for editing
2) yes - use mysql to default to now() if not entered - ie  ``<default>CURRENT_TIMESTAMP</default>`` - it's probably clearer on the form to also default to now & make the field required so people don't think they are saving 'empty' and then it isn't
3) yes - your update script looks correct.

Normally we try to be standardised about having both modified_date and created_date where one is present. In this case modified_date is present & we should probably add created_date for consistency - but since you are not in the end touching either it's OK not to consider that in scope - the xml for the modified_date & created_date should look like the below (when both are present)


```
 <field>
    <name>created_date</name>
    <type>timestamp</type>
    <comment>When the search was created.</comment>
    <required>true</required>
    <default>CURRENT_TIMESTAMP</default>
    <add>5.36</add>
  </field>
  <field>
    <name>modified_date</name>
    <type>timestamp</type>
    <comment>When the search was last modified.</comment>
    <required>true</required>
    <default>CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP</default>
    <add>5.36</add>
    <readonly>true</readonly>
  </field>
```

**Comment by jitendrapurohit** [comment]
Based on the discussion with @eileenmcnaughton the latest update to the PR does the following -

- Add `note_date` and `created_date` column to the note table.
- Both these columns are of type = timestamp (default to CURRENT_TIMESTAMP).
- Note date is exposed to UI and can be modified by the user.
- Upgrade handling - so the existing value of `modified_date` is copied to both the date columns.
- Test fix to assert the values in both the dates on create + update.

File: CRM/Core/DAO/Note.php [hunk 1]
```diff
@@ -220,11 +234,43 @@ public static function &fields() {
...
+          'localizable' => 0,
+          'add' => '5.36',
+        ],
         'modified_date' => [
           'name' => 'modified_date',
           'type' => CRM_Utils_Type::T_TIMESTAMP,
-          'title' => ts('Note Modified By'),
+          'title' => ts('Modified Date'),
           'description' => ts('When was this note last modified/edited'),
+          'required' => TRUE,
```
Comment (eileenmcnaughton) [inline CRM/Core/DAO/Note.php:273]: I'm not sure about adding required here - that would impact what needs to be submitted at the apiv4 level I think (@colemanw to confirm)

I think it's OK just relying on the default

**Comment by eileenmcnaughton** [comment]
The code looks good - just needs r-run testing & clarification on the required

**Comment by eileenmcnaughton** [comment]
@jitendrapurohit I was just thinking I should merge this before 5.37 is cut & then do the one fix that stalled this (https://github.com/civicrm/civicrm-core/pull/19738#pullrequestreview-616849883) as a follow up - but I discovered this is actually stale

**Comment by jitendrapurohit** [comment]
@eileenmcnaughton Fixed the rebase conflicts and removed the required attribute for `modified_date` field 👍

**Comment by eileenmcnaughton** [comment]
OK - that looks like the same code I tested minus the required - merging

**Comment by J0WI** [comment]
Hi there,

With MariaDB 10.4 this breaks with the error: `Unknown column 'note_date' in 'field list'`. Maybo some quotes or quotations or backticks are missing?

**Comment by eileenmcnaughton** [comment]
@J0WI under what circumstance does it break? It is possibly on upgrade because it is trying to access the field before it exists?

**Comment by J0WI** [comment]
At this point they already exist. I get the same error when I execute this manually:

```sql
ALTER TABLE civicrm_note ADD COLUMN `note_date` timestamp NULL  DEFAULT CURRENT_TIMESTAMP COMMENT 'Date attached to the note';
ALTER TABLE civicrm_note ADD COLUMN `created_date` timestamp NULL  DEFAULT CURRENT_TIMESTAMP COMMENT 'When the note was created';
UPDATE civicrm_note SET note_date = modified_date, created_date = modified_date, modified_date = modified_date;
```

**Comment by eileenmcnaughton** [comment]
So there are 3 statements there - do the first 2 run OK & not the last one?

**Comment by J0WI** [comment]
>  do the first 2 run OK & not the last one?

Exactly.

**Comment by eileenmcnaughton** [comment]
@J0WI - do you have data base logging enabled? ie is there  a log_civicrm_note table?

**Comment by J0WI** [comment]
Yes

**Comment by eileenmcnaughton** [comment]
@J0WI ok - so it will be the log_civicrm_note table that is missing the new field - if you can disable logging & re-enable it should create it

**Comment by J0WI** [comment]
Can this be included/fixed in the upgrade procedure?

**Comment by eileenmcnaughton** [comment]
@J0WI I found someone else had reported what sounded like the same issue https://lab.civicrm.org/dev/core/-/issues/2961 - I put up a patch here https://github.com/civicrm/civicrm-core/pull/22150 - which I think would fix it - but I haven't tested the patch

**Pull request merged**